source/vibe/core/sync.d

```diff
@@ -819,3 +819,490 @@ private struct TaskConditionImpl(bool INTERRUPTIBLE, LOCKABLE) {
+            //If the counter hits zero, notify waiting Tasks
+            activeCounter!INTENT -= 1;
+            if (activeCounter!INTENT == 0)
+            {
+                version(RWMutexPrint)
+                    printInfo!("notify",INTENT)();
+                
+                notify!INTENT();
+            }
+        }
+    }
+}
+
+/** A ReadWriteMutex implementation for fibers.
+ *
+ *  This mutex can be used in exchange for a $(D core.sync.mutex.ReadWriteMutex),
+ *  but does not block the event loop in contention situations. The `reader` and `writer`
+ *  members are used for locking. Locking the `reader` mutex allows access to multiple 
+ *  readers at once, while the `writer` mutex only allows a single writer to lock it at
+ *  any given time. Locks on `reader` and `writer` are mutually exclusive (i.e. whenever a 
+ *  writer is active, no readers can be active at the same time, and vice versa).
+ * 
+ *  The class is implemented as a template with an empty parameter list 
+ *  to allow the automatical inference of shared members. To instantiate it,
+ *  use the alias $(D TaskReadWriteMutex) defined in this module.
+ * 
+ *  Notice:
+ *      Mutexes implemented by this class cannot be interrupted
+ *      using $(D vibe.core.task.Task.interrupt()). The corresponding
+ *      InterruptException will be deferred until the next blocking
+ *      operation yields the event loop.
+ * 
+ *      Use $(D InterruptibleTaskReadWriteMutex) as an alternative that can be
+ *      interrupted.
+ * 
+ *  cf. $(D core.sync.mutex.ReadWriteMutex)
+ */
+class TaskReadWriteMutexImpl()
+{
+    private {
+        alias State = ReadWriteMutexState!false;
+        alias LockingIntent = State.LockingIntent;
+        alias READ_ONLY  = LockingIntent.READ_ONLY;
+        alias READ_WRITE = LockingIntent.READ_WRITE;
+        
+        /** The shared state used by the reader and writer mutexes. */
+        State m_state;
+    }
+    
+    /** The policy with which the mutex should operate.
+     *
+     *  The policy determines how the acquisition of the locks is 
+     *  performed and can be used to tune the mutex according to the
+     *  underlying algorithm in which it is used.
+     *
+     *  According to the provided policy, the mutex will either favor
+     *  reading or writing tasks and could potentially starve the 
+     *  respective opposite.
+     *
+     *  cf. $(D core.sync.rwmutex.ReadWriteMutex.Policy)
+     */
+    alias Policy = State.Policy;
+    
+    /** A common baseclass for both of the provided mutexes.
+     *
+     *  The intent for the according mutex is specified through the 
+     *  $(D INTENT) template argument, which determines if a mutex is 
+     *  used for read or write locking.
+     */
+    final class Mutex(LockingIntent INTENT): core.sync.mutex.Mutex, Lockable
+    {
+        /** Try to lock the mutex. cf. $(D core.sync.mutex.Mutex) */
+        override bool tryLock() { return m_state.tryLock!INTENT(); }
```

Comment:
Using a `std.core.sync.mutex.Mutex` (as suggested) leads to the non-interruptible Mutex implementation to also not being `nothrow` anymore.
It doesn't seem to be a real downside though, as I can't really see any sense in using a `TaskReadWriteMutex` as an object monitor anyway...